File: Medium/79. Word Search/solution.java
class Solution {
    int m, n;
    int[][] di = { { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 } };

    public boolean exist(char[][] board, String word) {
        m = board.length;
        n = board[0].length;
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                visited[i][j] = false;
            }
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (helper(board, word, 0, visited, i, j)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean helper(char[][] board, String word, int index, boolean[][] visited, int startx, int starty) {
        if (index == word.length() - 1) {
            return board[startx][starty] == word.charAt(index);
        }
        if (board[startx][starty] == word.charAt(index)) {
            visited[startx][starty] = true;
            for (int i = 0; i < 4; i++) {
                int newx = startx + di[i][0];
                int newy = starty + di[i][1];
                if (inArea(newx, newy) && !visited[newx][newy]) {
                    if (helper(board, word, index + 1, visited, newx, newy)) {
                        return true;
                    }
                }
            }
            visited[startx][starty] = false;
        }
        return false;
    }

    private boolean inArea(int x, int y) {
        return x >= 0 && x < m && y >= 0 && y < n;
    }
}

File: Medium/79. Word Search/solution.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        rows = len(board)
        cols = len(board[0])
        for i in range(rows):
            for j in range(cols):
                if self.dfs(board, word, i, j, 0):
                    return True
        return False
    
    def dfs(self, board: List[List[str]], word: str, x: int, y: int, index: int) -> bool:
        if index == len(word):
            return True
        rows = len(board)
        cols = len(board[0])
        if x >= 0 and x < rows and y >= 0 and y < cols:
            if board[x][y] != word[index]:
                return False
            prev = board[x][y]
            board[x][y] = "*"
            result = self.dfs(board, word, x +1, y, index + 1) or self.dfs(board, word, x - 1, y, index + 1) or self.dfs(board, word, x, y + 1, index + 1) or self.dfs(board, word, x, y - 1, index + 1)
            board[x][y] = prev
            return result
        else:
            return False
        